src/dali/plugin/pair_converter/ccxt.py

```diff
@@ -465,7 +550,39 @@ def _add_alternative_markets(self, graph: MappedGraph[str], current_markets: Dic
             # If it does exist it will look it up in the dictionary by name and add the neighbor to that vertex.
             graph.add_neighbor(base_asset, quote_asset, _ALTERNATIVE_MARKET_WEIGHT)
 
-    def _add_exchange_to_memcache(self, exchange: str) -> None:
+    def _cache_graph_snapshots(self, exchange: str) -> None:
+        # TO BE IMPLEMENTED - If asset is missing from manifest, warn user and reoptimize.
+        if self.__exchange_datetime_graph_tree_dict.get(exchange):
+            raise RP2ValueError(
+                f"Internal Error: You have already generated graph snapshots for exchange - {exchange}. " f"Optimization can only be performed once."
+            )
+
+        if self.__manifest:
+            unoptimized_graph: MappedGraph[str] = self._generate_unoptimized_graph(exchange)
+        else:
+            # TO BE IMPLEMENTED - Set a default start time of the earliest possible crypto trade.
+            raise RP2ValueError("INTERNAL ERROR: No manifest provided for the CCXT pair converter plugin. Unable to optimize the graph.")
+
+        optimizations: Dict[datetime, Dict[str, Dict[str, float]]]
+        optimizations = self._optimize_assets_for_exchange(
+            unoptimized_graph,
+            self.__manifest.first_transaction_datetime,
+            self.__manifest.assets,
+            exchange,
+        )
+        self.__logger.debug("Optimizations created for graph: %s", optimizations)
+
+        exchange_tree: AVLTree[datetime, MappedGraph[str]] = AVLTree[datetime, MappedGraph[str]]()
+        for timestamp, optimization in optimizations.items():
```

Comment:
My understanding of this loop is that there can be only one optimization per graph at a given timestamp. Is this correct/intended? Can there be a case in which a graph needs two or more optimizations at a given timestamp?

Reply:
I'm not sure if this was answered earlier, but yes there can be multiple optimizations at a given timestamp. I've done a little more optimization to try to reduce the number of snapshots made. Each snapshot will have all the optimizations of the previous snapshot, but then be updated with any changes, subtractions or additions.